cmd/frontend/graphqlbackend/users_randomize_password.go

```diff
@@ -2,53 +2,86 @@ package graphqlbackend
 
 import (
 	"context"
+	"net/url"
 
+	"github.com/cockroachdb/errors"
 	"github.com/graph-gophers/graphql-go"
 
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/backend"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/envvar"
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/globals"
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/internal/auth/userpasswd"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
 	"github.com/sourcegraph/sourcegraph/internal/database"
 	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
 )
 
 type randomizeUserPasswordResult struct {
-	db     dbutil.DB
-	userID int32
+	resetURL *url.URL
 }
 
-func (r *randomizeUserPasswordResult) ResetPasswordURL(ctx context.Context) (*string, error) {
-	if !userpasswd.ResetPasswordEnabled() {
-		return nil, nil
+func (r *randomizeUserPasswordResult) ResetPasswordURL() *string {
+	urlStr := globals.ExternalURL().ResolveReference(r.resetURL).String()
+	return &urlStr
+}
+
+func sendEmail(ctx context.Context, db dbutil.DB, userID int32, resetURL *url.URL) error {
+	user, err := database.Users(db).GetByID(ctx, userID)
+	if err != nil {
+		return err
 	}
 
-	// This method modifies the DB, which is somewhat counterintuitive for a "value" type from an
-	// implementation POV. Its behavior is justified because it is convenient and intuitive from the
-	// POV of the API consumer.
-	resetURL, err := backend.MakePasswordResetURL(ctx, r.db, r.userID)
+	email, _, err := database.UserEmails(db).GetPrimaryEmail(ctx, userID)
 	if err != nil {
-		return nil, err
+		return err
+	}
+
+	if err = userpasswd.SendResetPasswordURLEmail(ctx, email, user.Username, resetURL); err != nil {
+		return err
 	}
-	urlStr := globals.ExternalURL().ResolveReference(resetURL).String()
-	return &urlStr, nil
+
+	return nil
 }
 
 func (r *schemaResolver) RandomizeUserPassword(ctx context.Context, args *struct {
 	User graphql.ID
 }) (*randomizeUserPasswordResult, error) {
+	if !userpasswd.ResetPasswordEnabled() {
+		return nil, errors.New("Resetting passwords is not enabled")
+	}
+	if envvar.SourcegraphDotComMode() && !conf.CanSendEmail() {
+		return nil, errors.New("Unable to reset password because email sending is not configured")
+	}
 	// 🚨 SECURITY: Only site admins can randomize user passwords.
 	if err := backend.CheckCurrentUserIsSiteAdmin(ctx, r.db); err != nil {
 		return nil, err
 	}
 
 	userID, err := UnmarshalUserID(args.User)
 	if err != nil {
-		return nil, err
+		return nil, errors.New("Cannot parse user ID")
```

Comment:
```suggestion
		return nil, errors.Wrap(err, "cannot parse user ID")
```

The underlying error is swallowed.